azsettings/env.go

```diff
@@ -0,0 +1,55 @@
+package azsettings
+
+import (
+	"fmt"
+
+	"github.com/grafana/grafana-azure-sdk-go/azsettings/internal/envutil"
+)
+
+const (
+	envAzureCloud              = "GF_PLUGIN_AZURE_CLOUD"
```

Comment:
We use the `GF_PLUGIN_` prefix as a convention where Grafana auto creates an ENV vars based on plugin settings provided via the `ini`.

For example:

```ini
[plugin.grafana-azure-monitor-datasource]
cloud = true
```

becomes `export GF_PLUGIN_GRAFANA_AZURE_MONITOR_DATASOURCE_CLOUD=true`, which is subsequently accessible via a plugin. Further docs - see [here](https://grafana.com/docs/grafana/latest/administration/configuration/#override-configuration-with-environment-variables). 

For the AWS SDK, it's not using that mechanism. Instead it looks like so:
- [Grafana manually setting the env vars based on keys exposed in the AWS SDK](https://github.com/grafana/grafana/blob/main/pkg/setting/setting.go#L1095-L1103)
- [The AWS SDK reading those variables ](https://github.com/grafana/grafana-aws-sdk/blob/main/pkg/awsds/sessions.go#L53-L76)

Unless I'm forgetting something, the former should suit your cases - which would mean this line should be

```go
envAzureCloud = "GF_PLUGIN_{CORRECT_PLUGIN_ID}_CLOUD"
```

Reply:
If user set some settings for the host (via config or env vars), I would not want these raw settings to be automatically transferred to a plugin as is.

`AzureSettings` structure is very simple now but it will grow in complexity in future and will have state which is not direct user input but calculation of combinations of different user inputs and hosting environment (like combination of OAuth settings with Azure settings etc.).

Settings provided by a user are being validated/sanitized by the host ([/pkg/setting/setting_azure.go](https://github.com/grafana/grafana/blob/897db011eb6bde9c6cb57a1a1f0a96d06a7ba35b/pkg/setting/setting_azure.go#L5) - it is simple now) and resulting `AzureSettings` being used by built-in Azure datasources too. 

So I want transfer of the computed `AzureSettings` structure between host and plugin. This means serialization/deserialization of `AzureSettings` via env vars should be isolated from other effects.

I will make prefix unique for this package `WriteToEnvStr`/`ReadFromEnv` to `GFAZPL_` so it's guaranteed to not have another meaning in Grafana like (`GF_PLUGIN_` does):
```
GFAZPL_AZURE_CLOUD=
GFAZPL_AZURE_MANAGED_IDENTITY_ENABLED=
GFAZPL_AZURE_MANAGED_IDENTITY_CLIENT_ID=
```
I could even use UUIDs but it would complicate troubleshooting.

Reply:
> Settings provided by a user are being validated/sanitized by the host ([/pkg/setting/setting_azure.go](https://github.com/grafana/grafana/blob/897db011eb6bde9c6cb57a1a1f0a96d06a7ba35b/pkg/setting/setting_azure.go#L5) - it is simple now) and resulting `AzureSettings` being used by built-in Azure datasources too.

> So I want transfer of the computed `AzureSettings` structure between host and plugin. This means serialization/deserialization of `AzureSettings` via env vars should be isolated from other effects.

I don't really understand this, for example, since `cloud` is already a [setting in the configuration
](https://github.com/grafana/grafana/blob/main/conf/defaults.ini#L631), it can be set using a custom config file, and the environment variable `GF_AZURE_CLOUD`. Would that be different if the user set  `GFAZPL_AZURE_CLOUD` (or `AZURE_CLOUD` since that seems to be another alternative)? Is the goal to remove that configuration?

Reply:
Use can set cloud via `GF_AZURE_CLOUD` or via config file and this value goes through validation [here](https://github.com/grafana/grafana/blob/897db011eb6bde9c6cb57a1a1f0a96d06a7ba35b/pkg/setting/setting_azure.go#L12).

So the value stored in `cfg.Azure.Cloud` can be different from what user initially provided.

Built-in plugins won't see user's original value but the final value which is stored in `cfg.Azure.Cloud` (i.e. `azsettings.AzureSettings`).

When transferring contents of the `azsettings.AzureSettings` structure from host to external plugins, I want them to not be unintentionally mixed with other environment variables set on the machine. So I want the env keys to be unique enough. So, external plugins would received the same contents of `azsettings.AzureSettings` as it is on host.

Reply:
ah, I get it now, the problem is that plugins don't have access to the `cfg` information so the mechanism to allow this is to expose certain information as environment variables. The goal is for grafana/grafana to expose `GFAZPL_AZURE_CLOUD` so this SDK can read it, right? (not for the user to set that variable, that's why the name should be unique enough).